### PR TITLE
[DPE-1769] Minor version upgrades

### PR DESCRIPTION
**Pull request by marceloneppel**
## Issue
We cannot safely upgrade the charm.

## Solution
Implemented upgrade module with unit test coverage using the library from https://github.com/canonical/data-platform-libs/pull/77.

Due to https://github.com/pgbackrest/pgbackrest/issues/2007, there is no info about running backups in the json output from `pgbackrest info`, so we provided a workaround for that in `src/backups.py`, which sets a Patroni tag to inform that to the other cluster members (which is checked in `src/patroni.py`).

When the unit 0 is not a sync-standby (which is needed for a switchover when synchronous mode is enabled on Patroni), the charm changes some Patroni tags in the cluster members to turn unit zero a sync-standby and allow a switchover to it before the upgrade is started.

### Manual Testing (with same local built charm)
```shell
git clone git@github.com:canonical/postgresql-k8s-operator.git
cd postgresql-k8s-operator
git checkout dpe-1769-minor-version-upgrade

charmcraft pack
# Deploy local charm
juju deploy ./*.charm --trust --resource postgresql-image=ghcr.io/canonical/charmed-postgresql@sha256:2181002bffd2e8aa7ddf3e5584629b3fa5ac47ab7fa0b1627bc2e093331bbd59 -n 3

# After units settle, run pre-upgrade-check to build update-stack.
# This will change primary to unit 0 if not already (if unit 0 is not a sync-standby, we cannot switchover to it, so this actions perform changes in the cluster member roles to enable the switchover in the next action run - after 30 seconds between one run and the second).
juju run-action postgresql-k8s/leader pre-upgrade-check --wait

# Refresh command - we are updating to the same charm, which will pass version checking.
juju refresh --path ./*.charm postgresql-k8s --resource postgresql-image=ghcr.io/canonical/charmed-postgresql@sha256:2181002bffd2e8aa7ddf3e5584629b3fa5ac47ab7fa0b1627bc2e093331bbd59

# At this point just the highest ordinal unit will be updated.
# After successful upgrade, run:
juju run-action postgresql-k8s/leader resume-upgrade --wait

# Which will trigger auto rollout for upgrading the other units.
```

While reviewing, the following files can be ignored (they are either imported or generated):
* `lib/charms/data_platform_libs/v0/upgrade.py`
* `requirements.txt`
* `tests/integration/ha_tests/application-charm/requirements.txt`
* `tests/integration/new_relations/application-charm/requirements.txt`

We disabled the hashes temporarily on all the `requirements.txt` files due to [canonical/charmcraft#1179] (so we can pack the charms now). I will be reverted later.

### CODE REVIEW

**Review by delgod — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 1]
```diff
@@ -1,442 +1,35 @@
-anyio==3.7.1 ; python_full_version >= "3.10.6" and python_full_version < "4.0.0" \
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply requirements.txt:1]: Yes, we can. But we disabled it temporarily due to https://github.com/canonical/charmcraft/issues/1179. I forgot to add it to this PR description. I added it now. We used `poetry export --without-hashes -f requirements.txt -o requirements.txt` to make the charm build until the issue in charmcraft is resolved.

File: src/upgrade.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            logger.debug("Deferring on_pebble_ready: no upgrade peer relation yet")
+            event.defer()
+            return
+
+        if self.peer_relation.data[self.charm.unit].get("state") != "upgrading":
+            return
+
+        # Don't mark the upgrade of this unit as completed until Patroni reports the
+        # workload is ready.
+        if not self.charm._patroni.member_started:
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/upgrade.py:89]: I think it's possible. I'm gonna implement it today.

File: src/dependency.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/dependency.json:1]: I'm not sure if it's possible or not. Let me check and test it.

[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/dependency.json:1]: It's possible. I specified the rock PG version in https://github.com/canonical/postgresql-k8s-operator/pull/199/commits/1f246e2f360067876632c223223c3806a95d3582.

File: src/upgrade.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/upgrade.py:88]: Added on https://github.com/canonical/postgresql-k8s-operator/pull/199/commits/1f246e2f360067876632c223223c3806a95d3582.

**Review by delgod — APPROVED** [review]

**Review by dragomirp — APPROVED** [review]

**Pull request merged**